### PR TITLE
Add auditing to feature flags

### PR DESCRIPTION
**Pull request by scottbommarito**
Gallery side portion of https://github.com/nuget/engineering/issues/2333

### CODE REVIEW

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using NuGet.Services.FeatureFlags;
+
+namespace NuGetGallery.Auditing.AuditedEntities
+{
+    public class AuditedFeatureFlagFlight
+    {
+        public string Name { get; private set; }
+        public bool All { get; private set; }
+        public bool SiteAdmins { get; private set; }
+        public int AccountsCount { get; private set; }
```
Comment (scottbommarito) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: Note that I'm auditing the number of accounts and domains in the flight rather than the actual values--I didn't want to put extra effort into making the account/domain names GDPR compliant.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFeature.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        public string Name { get; private set; }
+        public FeatureStatus Status { get; private set; }
+
+        public static AuditedFeatureFlagFeature[] CreateFrom(FeatureFlags flags)
+        {
+            return flags.Features
+                .Select(f => CreateFrom(f.Key, f.Value))
+                .ToArray();
+        }
+        public static AuditedFeatureFlagFeature CreateFrom(string name, FeatureStatus status)
```
Comment (loic-sharma) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFeature.cs]: Please add a new line above here

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: @cristinamanum We can put PII in auditing data as it's ephemeral, right?

File: src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs [hunk 3]
```diff
@@ -118,6 +121,19 @@ public async Task RemoveUserAsync(User user)
         }
 
         public async Task<FeatureFlagSaveResult> TrySaveAsync(FeatureFlags flags, string contentId)
+        {
+            var result = await TrySaveInternalAsync(flags, contentId);
+            await _auditing.SaveAuditRecordAsync(
```
Comment (loic-sharma) [inline src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs]: If there's an exception in between auditing and saving the flags, what data would we rather lose? It seems to me that we should guarantee that all update attempts are audited.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: I noticed that some of the other auditing classes have `Obfuscate` attributes on them. My problem with having an array of usernames in this audit record is that this obfuscation attribute doesn't currently support arrays.

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            if (flags == null)
+            {
+                throw new ArgumentNullException(nameof(flags));
+            }
+
+            Features = AuditedFeatureFlagFeature.CreateFrom(flags);
+            Flights = AuditedFeatureFlagFlight.CreateFrom(flags);
+
+            ContentId = contentId ?? throw new ArgumentNullException(nameof(contentId));
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: move this validation to the beginning of the ctor (or right after flags check)

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFeature.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+namespace NuGetGallery.Auditing.AuditedEntities
+{
+    public class AuditedFeatureFlagFeature
+    {
+        public string Name { get; private set; }
+        public FeatureStatus Status { get; private set; }
+
+        public static AuditedFeatureFlagFeature[] CreateFrom(FeatureFlags flags)
+        {
+            return flags.Features
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFeature.cs]: >Features [](start = 25, length = 8)

Features can be null

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (skofman1) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: yes

---
In reply to: [274024465](https://github.com/NuGet/NuGetGallery/pull/7058#discussion_r274024465) [](ancestors = 274024465)

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        public int DomainsCount { get; private set; }
+
+        public static AuditedFeatureFlagFlight[] CreateFrom(FeatureFlags flags)
+        {
+            return flags.Flights
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: this can be null

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (skofman1) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: it's ok to put usernames in audit, as long as we don't audit to blob storage but only to IFxAudit

---
In reply to: [273941956](https://github.com/NuGet/NuGetGallery/pull/7058#discussion_r273941956) [](ancestors = 273941956)

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+namespace NuGetGallery.Auditing
+{
+    public sealed class FeatureFlagsAuditRecord : AuditRecord<AuditedFeatureFlagsAction>
+    {
+        public AuditedFeatureFlagFeature[] Features { get; }
+        public AuditedFeatureFlagFlight[] Flights { get; }
+        public string ContentId { get; }
+        public FeatureFlagSaveResult Result { get; }
+
+        public FeatureFlagsAuditRecord(
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: we also want to know who was the user that performed the update

Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: I was under the impression that the "audit actor", which is automatically injected into the audit log, contains this information. I will double check and add it if not.

Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: I was under the impression that the "audit actor", which is automatically injected into the audit log, contains this information. I will double check and add it if not.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: Will add a new obfuscation type (`UserNameList`) that obfuscates as `new [] { Obfuscator.ObfuscatedUserName }` and then include the full list of usernames in the audit record.

File: src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs]: I'm not sure. AI will already have record of these attempts and the exceptions they throw. Many bad requests will fail before we even get to save the flags--should we log those too? I was under the impression that these auditing changes were more for seeing the change history and the user that made the change. I would argue that failed changes are much less important for this.

Furthermore, I looked across the gallery. It seems we're very inconsistent regarding whether or not to audit the change before or after the actual change occurs, but I don't see any cases in which we try-catch-finally and always audit the change.

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs (hunk 7 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: In the JSON blob, this is the case. I will double-check on IFXAudit before merging.

File: src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscatorJsonConverter.cs [hunk 8]
```diff
@@ -41,10 +44,30 @@ protected virtual PropertyInfo ResolvePath(string path)
             var currentInstance = Instance;
             foreach (var part in path.Split('.'))
             {
-                property = currentInstance.GetType().GetProperty(part);
-                if (property == null) { return null; }
+                var propertyName = part;
```
Comment (scottbommarito) [inline src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscatorJsonConverter.cs]: Obfuscation was not being done for fields of objects that were in an array.

For example:
```
{
    hello: [
        {
            "someField": "this is not being obfuscated correctly!"
        }
    ]
}
```

This change fixes it, but ultimately I think this entire obfuscation converter should be rewritten in a way that doesn't require resolving the path of the object being serialized. There are a lot of scenarios here that are untested--dictionaries, arrays with multiple indices, etc.

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            ContentId = contentId ?? throw new ArgumentNullException(nameof(contentId));
+
+            Features = AuditedFeatureFlagFeature.CreateFrom(flags);
+            Flights = AuditedFeatureFlagFlight.CreateFrom(flags);
+            
+            Result = result;
+
+            // Group feature flags changes by the month they occurred in.
+            var currentTime = DateTime.UtcNow;
+            _path = $"{currentTime.Year}-{currentTime.Month}";
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: retention policy for IFxAudit is 1 week, so this will not be useful.  You might as well use a const value.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+using NuGetGallery.Auditing.Obfuscation;
+
+namespace NuGetGallery.Auditing.AuditedEntities
+{
+    public class AuditedFeatureFlagFlight
+    {
+        public string Name { get; private set; }
+        public bool All { get; private set; }
+        public bool SiteAdmins { get; private set; }
+        [Obfuscate(ObfuscationType.UserNameList)]
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: >i [](start = 27, length = 1)

I'm confused: this audit should go ONLY to IfxAudit. And there we don't need to obfuscate. So why is this needed?

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (cristinamanum) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: @loic-sharma and @skofman1 are correct, we do not need to obfuscate the data that goes audit logs.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 10 above)
Comment (cristinamanum) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: We need to obfuscate only if this data is persisted to blob storage or anything else than ifx audit.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public static AuditedFeatureFlagFlight CreateFrom(string name, Flight flight)
+        {
+            return new AuditedFeatureFlagFlight
+            {
+                Name = name,
+                All = flight.All,
+                SiteAdmins = flight.SiteAdmins,
+                Accounts = flight.Accounts?.ToArray() ?? new string[0],
```
Comment (cristinamanum) [inline src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: This could mean a large number of accounts. Maybe is worth a manual testing for how it behaves for large number of accounts.

File: src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs [hunk 12]
```diff
@@ -8,6 +8,7 @@ public enum ObfuscationType
         UserKey,
         UserName,
         Authors,
-        IP
+        IP,
+        UserNameList
```
Comment (cristinamanum) [inline src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs]: If the logs go only in IFXAudit these changes are not needed right?

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 10 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: There are three implementations of the auditing service--file, cloud, and IFX. Each implementation decides whether or not to obfuscate the data, and only cloud does it today. There is no mechanism for saying "this should only be audited if we are doing IFX audit". As a result, we need to make sure it has the proper obfuscation attribute in case we are not using the IFX audit implementation.

As a side note, cloud is the only auditing service implementation that processes the obfuscation attribute--the implementation that writes to the local file system does not obfuscate. As it is meant to be a debugging tool it seems fine to me but I thought it was worth mentioning.

File: src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs (hunk 3 above)
Comment (cristinamanum) [reply src/NuGetGallery.Core/Features/FeatureFlagFileStorageService.cs]: I think that the TrySaveInternalAsync exception data is more of diagnostics data rather than Audit data.

File: src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs (hunk 12 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs]: There is no mechanism today to prevent them from going to another source other than IFX audit. I would rather not break the pattern of the existing code--which is to use the `Obfuscate` attribute on any audit property that should be obfuscated.

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs (hunk 9 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: I believe this is mostly used by the local file system auditing service. It uses this string to determine where to save the audit record file.

**Comment by scottbommarito** [comment]
@cristinamanum - yup, shims changes coming soon

**Comment by scottbommarito** [comment]
@skofman1 @cristinamanum - it seems like I was wrong and that there is a mechanism that prevents any audit logs other than package deletion audit records from going to blob storage, so I can go ahead and remove all obfuscation changes from this PR

File: src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs (hunk 7 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/FeatureFlagsAuditRecord.cs]: It is the case on IFXAudit.

File: src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs (hunk 12 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscationType.cs]: Determined that I was wrong.

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 10 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: This log will not go to blob storage, so no obfuscation is needed.

File: src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscatorJsonConverter.cs (hunk 8 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/Obfuscation/ObfuscatorJsonConverter.cs]: no longer necessary

**Comment by scottbommarito** [comment]
Shims changes at https://nuget.visualstudio.com/NuGetMicrosoft/_git/NuGet.Shims/pullrequest/899?_a=files

File: src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedFeatureFlagFlight.cs]: no need to obfuscate. putting full names in audit logs
